tests/unit/test_req_file.py

```diff
@@ -382,6 +382,10 @@ def test_set_finder_allow_all_prereleases(self, line_processor, finder):
         line_processor("--pre", "file", 1, finder=finder)
         assert finder.allow_all_prereleases
 
+    def test_use_feature(self, line_processor):
+        """--use-feature can be set in requirements files."""
+        line_processor("--use-feature=2020-resolver", "filename", 1)
```

Comment:
```suggestion
        line_processor("--use-feature=2020-resolver", "filename", 1, options=options)
        assert "2020-resolver" in options.features_enabled
```
The `options` fixture could be used to assert whether the feature is on.

And also, IIUC, `options.features_enabled` should be updated here (same as `options.require_hashes`): https://github.com/pypa/pip/blob/982aac544ecfbbb3b18d87fec51ed8ce3d957494/src/pip/_internal/req/req_file.py#L227-L229.

Reply:
> The `options` fixture could be used to assert whether the feature is on.

Ah nice. I was just following the pattern of the other tests in the file. I'll fix and take a look at the other location

Reply:
Actually nevermind. The `options` fixture doesn't get updated here.

> And also, perhaps, options.features_enabled should be updated here (same as options.require_hashes):

It's not necessary, this already works:

```console
$ cat a.txt 
-r b.txt
$ cat b.txt 
boto3
boto3==1.13.13
$ pip install -r a.txt 
Collecting boto3
  Using cached boto3-1.13.13-py2.py3-none-any.whl (128 kB)
Collecting s3transfer<0.4.0,>=0.3.0
  Using cached s3transfer-0.3.3-py2.py3-none-any.whl (69 kB)
Collecting jmespath<1.0.0,>=0.7.1
  Using cached jmespath-0.10.0-py2.py3-none-any.whl (24 kB)
Collecting botocore<1.17.0,>=1.16.13
  Using cached botocore-1.16.26-py2.py3-none-any.whl (6.2 MB)
Collecting docutils<0.16,>=0.10
  Using cached docutils-0.15.2-py3-none-any.whl (547 kB)
Collecting urllib3<1.26,>=1.20
  Using cached urllib3-1.25.10-py2.py3-none-any.whl (127 kB)
Collecting python-dateutil<3.0.0,>=2.1
  Using cached python_dateutil-2.8.1-py2.py3-none-any.whl (227 kB)
Collecting six>=1.5
  Using cached six-1.15.0-py2.py3-none-any.whl (10 kB)
Installing collected packages: six, urllib3, python-dateutil, jmespath, docutils, botocore, s3transfer, boto3
Successfully installed boto3-1.13.13 botocore-1.16.26 docutils-0.15.2 jmespath-0.10.0 python-dateutil-2.8.1 s3transfer-0.3.3 six-1.15.0 urllib3-1.25.10
```

Reply:
If it works — it's great! The only point is there should be at least one assertion in the test. What do you think?

Reply:
I tested and it does work indeed. But I suspect the option is propagated accidentally because it is appended to the default value which is a [global list](https://github.com/pypa/pip/blob/50aca38eafde71ac0314d14246f7ebf043695f5e/src/pip/_internal/cli/cmdoptions.py#L915).

So to avoid additional headaches to whoever is going to refactor that req file parser, I'd suggest applying both changes as suggested by @atugushev (namely the assert and the option propagation in handle_option_line).

Would you mind doing that @rouge8? Otherwise it looks good to me and I can merge when the change is done.

Reply:
I was looking into the assert but I couldn't find anything to assert on and most of the other tests in that file don't have it either... Open to tips here

> But I suspect the option is propagated accidentally because it is appended to the default value which is a global list

I can take a look or I could add a more integration-y test to verify the behavior is maintained when someone tries to refactor that. Does that sound okay?

Reply:
> The only point is there should be at least one assertion in the test.

Arguably, the test is asserting that there are no errors while parsing the --use-feature option which implies that it's being accepted. I think that's a sufficient check, as long as we have a test to make sure we reject invalid options. :)

Reply:
@xavfernandez any suggestions on a more precise test?

Reply:
@rouge8 I propose https://github.com/rouge8/pip/pull/2

Reply:
ahhhh thanks! @sbidoul